daemon/snap.go

```diff
@@ -338,12 +342,68 @@ func fillComponentInfo(about aboutSnap) []client.Component {
 	return comps
 }
 
-// snapIcon tries to find the icon inside the snap
+// snapIcon tries to find the icon inside the snap.
 func snapIcon(info snap.PlaceInfo) string {
 	found, _ := filepath.Glob(filepath.Join(info.MountDir(), "meta", "gui", "icon.*"))
-	if len(found) == 0 {
+	if len(found) > 0 {
+		return found[0]
+	}
+
+	// Try to find the icon via the Icon= field of a .desktop file for the snap.
+
+	// Find all .desktop files in ${SNAP}/meta/gui
+	desktopFiles, _ := filepath.Glob(filepath.Join(info.MountDir(), "meta", "gui", "*.desktop"))
+	if len(desktopFiles) == 0 {
 		return ""
 	}
 
-	return found[0]
+	// Try <snap>.desktop first
+	if snapDesktop := info.SnapName() + ".desktop"; strutil.ListContains(desktopFiles, snapDesktop) {
+		if iconPath, err := iconPathFromDesktopFile(info, snapDesktop); err == nil {
+			return iconPath
+		}
+	}
+
+	// Try <snap>_<snap>.desktop next
+	if snapAppDesktop := fmt.Sprintf("%s_%s.desktop", info.SnapName(), info.SnapName()); strutil.ListContains(desktopFiles, snapAppDesktop) {
+		if iconPath, err := iconPathFromDesktopFile(info, snapAppDesktop); err == nil {
+			return iconPath
+		}
+	}
+
+	// Try any .desktop file we could find
+	for _, desktopFile := range desktopFiles {
+		if iconPath, err := iconPathFromDesktopFile(info, desktopFile); err == nil {
+			return iconPath
+		}
+	}
+
+	return ""
+}
+
+var iconMatcher = regexp.MustCompile(fmt.Sprintf("^%s(.*)$", regexp.QuoteMeta("Icon=${SNAP}")))
```

Comment:
For desktop files in `/snap` this works for the vscode snap but fails for firefox:
```bash
$ cat /snap/code/current/meta/gui/code.desktop | grep 'Icon'
Icon=${SNAP}/meta/gui/vscode.png
Icon=${SNAP}/meta/gui/vscode.png

$ cat /snap/firefox/current/firefox.desktop | grep 'Icon'
Icon=/default256.png
```

In the case of the files cat `/var/lib/snapd/desktop/applications` (what I ended up resolving in https://github.com/canonical/aa-prompting-test/pull/26) it looks like everything is an absolute path?
```bash
$ cat /var/lib/snapd/desktop/applications/code_code.desktop | grep 'Icon'
Icon=/snap/code/163/meta/gui/vscode.png
Icon=/snap/code/163/meta/gui/vscode.png

$ cat /var/lib/snapd/desktop/applications/firefox_firefox.desktop | grep 'Icon'
Icon=/snap/firefox/4483/default256.png
```